src/net_processing.cpp

```diff
@@ -2626,7 +2645,11 @@ void PeerManagerImpl::ProcessGetCFCheckPt(CNode& peer, CDataStream& vRecv)
 void PeerManagerImpl::ProcessBlock(CNode& node, const std::shared_ptr<const CBlock>& block, bool force_processing)
 {
     bool new_block{false};
-    m_chainman.ProcessNewBlock(block, force_processing, &new_block);
+    if (m_chainman.ProcessNewBlock(block, force_processing, &new_block)) {
+        LOCK(cs_main);
+        m_chainman.UpdateChainTipSet(block->GetHash(), GetChainTipSetForPeerNetwork(node));
```

Comment:
`UpdatedBlockTip()` does not convey any information about the origin (the peer that sent us the block) of the new tip which makes it impossible to update the tip set without storing that information separately. `mapBlockSource` would work for that, if we wouldn't remove elements from it in `BlockChecked`.

A slight improvement could be to move the `UpdateChainTipSet()` call a couple lines down into `if(new_block)`?

There are also two more calls to `UpdateChainTipSet()` (when receiving headers and compact blocks) that could not be moved into `UpdatedBlockTip()`, since `UpdatedBlockTip()` only fires for full blocks but we need to update the tip set if we only receive headers as well.